File: 1.php-pg/src/Service/RangeService.php
<?php
/**
 * Created by PhpStorm.
 * User: lenarsagiev
 * Date: 2019-02-14
 * Time: 18:07
 */

namespace Service;


use http\Exception;

class RangeService
{
    private $pdo;

    public function __construct(\PDO $pdo)
    {
        $this->pdo = $pdo;
    }

    public function createRange(int $min, int $max): int
    {
        if ($max < $min) {
            throw new \RuntimeException('Cannot create range with max_value less than min_value');
        }

        $stmt = $this->pdo->prepare('INSERT INTO "ranges"("start", "end", "free") VALUES(:start, :end, :free);');
        $stmt->execute([
            ':start' => $min,
            ':end' => $max,
            ':free' => $max - $min + 1
        ]);
        $rangeId = $this->pdo->lastInsertId();
        return $rangeId;
    }

    public function allocate(int $rangeId): int
    {
        $stmt = $this->pdo->prepare('SELECT allocate_number(:rangeid);');
        $stmt->execute([':rangeid' => $rangeId]);
        if ($stmt->errorCode() !== "00000") {
            throw new \RuntimeException('Cannot allocate more numbers');
        }
        return $stmt->fetchColumn(0);
    }

    public function release(int $rangeId, int $number): void
    {
        $stmt = $this->pdo->prepare('SELECT release_number(:rangeid, :number);');
        $stmt->execute([':rangeid' => $rangeId, ':number' => $number]);
        if ($stmt->errorCode() !== '00000') {
            throw new \RuntimeException('Cannot release number');
        }
    }
}

File: 2.jvm/src/main/java/service/RangeService.java
package service;

import model.Range;

import java.util.ArrayList;

public class RangeService {
    private ArrayList<Range> rangePool = new ArrayList<>();

    /**
     * Create new range and return its id.
     *
     * @param minValue Left edge of range.
     * @param maxValue Right edge of range.
     * @return index.
     */
    public int createRange(long minValue, long maxValue) {
        int index;
        synchronized (rangePool) {
            index = rangePool.size();
            rangePool.add(index, new Range(index, minValue, maxValue));
        }
        return index;
    }

    /**
     * Allocate number from range.
     *
     * @param rangeId id of range.
     * @return free number from range.
     */
    public long allocate(int rangeId) throws Exception {
        return getRange(rangeId).allocate();
    }

    /**
     * Release number from range.
     *
     * @param rangeId index of range.
     * @param number number to release.
     */
    public void release(int rangeId, long number) throws Exception {
        getRange(rangeId).release(number);
    }

    /**
     * Get range by its id, check if it's not correct.
     *
     * @param rangeId range index.
     * @return range model object.
     */
    private Range getRange(int rangeId) throws Exception {
        Range range = rangePool.get(rangeId);
        if (range == null) {
            throw new Exception("Incorrect range id");
        }
        return range;
    }
}
